core/src/main/java/com/linecorp/armeria/server/DefaultServiceRequestContext.java

```diff
@@ -434,82 +439,102 @@ public void setMaxRequestLength(long maxRequestLength) {
         final HttpHeaders additionalResponseTrailers = this.additionalResponseTrailers;
         if (additionalResponseTrailers == null) {
-            return HttpHeaders.EMPTY_HEADERS;
+            return HttpHeaders.of();
         }
-        return additionalResponseTrailers.asImmutable();
+        return additionalResponseTrailers;
     }
 
     @Override
-    public void setAdditionalResponseTrailer(AsciiString name, String value) {
+    public void setAdditionalResponseTrailer(CharSequence name, Object value) {
         requireNonNull(name, "name");
         requireNonNull(value, "value");
-        createAdditionalTrailersIfAbsent().set(name, value);
+        additionalResponseTrailers = createAdditionalTrailersIfAbsent().toBuilder()
+                                                                       .setObject(name, value).build();
     }
 
     @Override
-    public void setAdditionalResponseTrailers(Headers<? extends AsciiString, ? extends String, ?> headers) {
+    public void setAdditionalResponseTrailers(Iterable<? extends Entry<? extends CharSequence, ?>> headers) {
         requireNonNull(headers, "headers");
-        createAdditionalTrailersIfAbsent().set(headers);
+        additionalResponseTrailers = createAdditionalTrailersIfAbsent().toBuilder().setObject(headers).build();
     }
 
     @Override
-    public void addAdditionalResponseTrailer(AsciiString name, String value) {
+    public void addAdditionalResponseTrailer(CharSequence name, Object value) {
         requireNonNull(name, "name");
         requireNonNull(value, "value");
-        createAdditionalTrailersIfAbsent().add(name, value);
+        additionalResponseTrailers = createAdditionalTrailersIfAbsent().toBuilder()
+                                                                       .addObject(name, value).build();
     }
 
     @Override
-    public void addAdditionalResponseTrailers(Headers<? extends AsciiString, ? extends String, ?> headers) {
+    public void addAdditionalResponseTrailers(Iterable<? extends Entry<? extends CharSequence, ?>> headers) {
         requireNonNull(headers, "headers");
-        createAdditionalTrailersIfAbsent().add(headers);
+        additionalResponseTrailers = createAdditionalTrailersIfAbsent().toBuilder().addObject(headers).build();
     }
 
     @Override
-    public boolean removeAdditionalResponseTrailer(AsciiString name) {
+    public boolean removeAdditionalResponseTrailer(CharSequence name) {
         requireNonNull(name, "name");
-        return createAdditionalTrailersIfAbsent().remove(name);
+        final HttpHeaders additionalResponseTrailers = this.additionalResponseTrailers;
+        if (additionalResponseTrailers == null || additionalResponseTrailers.isEmpty()) {
+            return false;
+        }
+
+        final HttpHeadersBuilder builder = createAdditionalTrailersIfAbsent().toBuilder();
+        final boolean removed = builder.remove(name);
+        this.additionalResponseTrailers = builder.build();
```

Comment:
ditto

Reply:
Updated `*HeadersBuilder` so that `build()` returns the original headers if no changes were made.